src/companion.jl

```diff
@@ -15,86 +16,78 @@ Also, directly from a polynomial:
 ```julia
 julia> using Polynomials
 
-julia> P=Polynomial([2.0,3,4,5])
+julia> P = Polynomial(2:5)
 Polynomial(2 + 3x + 4x^2 + 5x^3)
 
-julia> C=Companion(P)
+julia> C = Companion(P)
 3×3 Companion{Float64}:
  0.0  0.0  -0.4
  1.0  0.0  -0.6
  0.0  1.0  -0.8
 ```
 """
-struct Companion{T} <: AbstractArray{T, 2}
+struct Companion{T} <: AbstractMatrix{T}
     c :: Vector{T}
 end
 
-# Generate companion matrix from a polynomial
+Companion(v::AbstractVector) = Companion(collect(v))
 
-function Companion(P::Polynomial{T}) where T
-   n = length(P)
-   c = Array{T}(undef, n-1)
-   d=P.coeffs[n]
-   for i=1:n-1
-       c[i]=P.coeffs[i]/d
-   end
-   Companion(c)
+# Construct companion matrix from a polynomial
+
+function Companion(P::Polynomial)
+    n = length(P)
+    c = P.coeffs[1:(n-1)] / P.coeffs[n]
+    return Companion(c)
 end
 
-#Basic property computations
-size(C::Companion, r::Int) = (r==1 || r==2) ? length(C.c) :
-    throw(ArgumentError("Companion is of rank 2"))
+# Basic properties
 
 function size(C::Companion)
     n = length(C.c)
-    n, n
+    return n, n
 end
 
-#XXX Inefficient but works
-# getindex(C::Companion, i, j) = getindex(Matrix(C), i, j)
-# isassigned(C::Companion, i, j) = isassigned(Matrix(C), i, j)
-getindex(C::Companion{T}, i::Int, j::Int) where T = (j==length(C.c)) ? -C.c[i] : (i==j+1 ? one(T) : zero(T) )
-isassigned(C::Companion, i::Int, j::Int) = (j==length(C.c)) ? isassigned(C.c,i) : true
```

Comment:
Why is `isassigned` being removed here?

Reply:
HI @mkitti, thank for the very thorough review.  I will address (or commit) your comments one-by-one, which might take a while, and I will ping you again when it is ready for another review.
This repo goes back to early versions of Julia and many things in the original version are no longer needed in current Julia versions.  Now `Base` has default fallbacks for `isassigned` for any `AbstractArray` types so there is no need for this function here.  To double check that that fallback is working correctly, I have added tests of this form:
```julia
@test isassigned(Z, 1)
@test isassigned(Z, n^2)
@test !isassigned(Z, 0)
@test !isassigned(Z, n^2+1)
```
Now, these are all standard uses of `isassigned` with a single index argument per the [manual](https://docs.julialang.org/en/v1/base/arrays/#Base.isassigned).  I have never seen a 2-index use of `isassigned` before, and it was never tested in this package.  If there is a use case for it, then it's fine to restore it, but then it should have a docstring and jldoctests that explains what it does.
Note that `SpecialMatrices.jl` does an `import` of `Base.isassigned` and this is the only one of the special matrices that extends this method, so it really looked like a bug (or historical baggage) to be importing the 1-index version and then defining a 2-index version.  Again, happy to restore it if there is a reason to have this non-standard method...

Reply:
Here is the fallback function.  While technically this works, I think we may be able to do this more efficiently.
```julia
julia> @which isassigned(Companion(1:3),1,1)
isassigned(a::AbstractArray, i::Integer...) in Base at abstractarray.jl:563

function isassigned(a::AbstractArray, i::Integer...)
    try
        a[i...]
        true
    catch e
        if isa(e, BoundsError) || isa(e, UndefRefError)
            return false
        else
            rethrow()
        end
    end
end
```

That said that previous definition does seem problematic.

```julia
julia> using SpecialMatrices

julia> C = Companion(1:3)
3×3 Companion{Int64}:
 0  0  -1
 1  0  -2
 0  1  -3

julia> isassigned(C, 1, 4)
false

julia> Base.isassigned(C::Companion, i::Int, j::Int) = (j==length(C.c)) ? isassigned(C.c,i) : true

julia> isassigned(C, 1, 4)
true
```

My question is does anything depend on this unusual behavior? Otherwise, I suggest we consider writing an optimized version of this.

Reply:
Similarly, using the old definition of `getindex` changes the situation as well.

```julia
julia> Base.getindex(C::Companion{T}, i::Int, j::Int) where T = (j==length(C.c)) ? -C.c[i] : (i==j+1 ? one(T) : zero(T) )

julia> C[1,4]
0
```

This is also weird. Just to be clear what we are doing here is materially changing the functionality of `Companion`. At the moment this seems odd, but perhaps there is a reason it is like this.

```julia
julia> [C[ci] for ci in CartesianIndices((3,10))]
3×10 Matrix{Int64}:
 0  0  -1  0  0  0  0  0  0  0
 1  0  -2  0  0  0  0  0  0  0
 0  1  -3  0  0  0  0  0  0  0
```

This seems amenable adding additional columns.